weed/server/volume_server_handlers.go

```diff
@@ -36,55 +36,76 @@ func (vs *VolumeServer) privateStoreHandler(w http.ResponseWriter, r *http.Reque
-			select {
-			case <-r.Context().Done():
-				glog.V(4).Infof("request cancelled from %s: %v", r.RemoteAddr, r.Context().Err())
-				return
-			default:
-				glog.V(4).Infof("wait because inflight download data %d > %d", vs.inFlightDownloadDataSize, vs.concurrentDownloadLimit)
-				vs.inFlightDownloadDataLimitCond.Wait()
+
+		inFlightDownloadSize := atomic.LoadInt64(&vs.inFlightDownloadDataSize)
+		if vs.concurrentDownloadLimit != 0 && inFlightDownloadSize > vs.concurrentDownloadLimit {
+			for {
+				glog.V(4).Infof("wait because inflight download data %d > %d", inFlightDownloadSize, vs.concurrentDownloadLimit)
+				select {
+				case <-r.Context().Done():
+					glog.V(4).Infof("request cancelled from %s: %v", r.RemoteAddr, r.Context().Err())
+					go func() {
+						<-vs.inFlightDownloadWaitCh
+					}()
+					return
+				case <-vs.inFlightDownloadWaitCh:
+				}
+
+				inFlightDownloadSize = atomic.LoadInt64(&vs.inFlightDownloadDataSize)
+				if inFlightDownloadSize <= vs.concurrentDownloadLimit {
+					break
+				}
 			}
 		}
-		vs.inFlightDownloadDataLimitCond.L.Unlock()
+
 		vs.GetOrHeadHandler(w, r)
+
 	case "DELETE":
 		stats.DeleteRequest()
 		vs.guard.WhiteList(vs.DeleteHandler)(w, r)
+
 	case "PUT", "POST":
 		contentLength := getContentLength(r)
 		// exclude the replication from the concurrentUploadLimitMB
 		if r.URL.Query().Get("type") != "replicate" && vs.concurrentUploadLimit != 0 {
 			startTime := time.Now()
-			vs.inFlightUploadDataLimitCond.L.Lock()
-			for vs.inFlightUploadDataSize > vs.concurrentUploadLimit {
+
+			for {
+				inFlightUploadSize := atomic.LoadInt64(&vs.inFlightUploadDataSize)
+				if inFlightUploadSize <= vs.concurrentUploadLimit {
+					break
+				}
+
 				//wait timeout check
 				if startTime.Add(vs.inflightUploadDataTimeout).Before(time.Now()) {
-					vs.inFlightUploadDataLimitCond.L.Unlock()
-					err := fmt.Errorf("reject because inflight upload data %d > %d, and wait timeout", vs.inFlightUploadDataSize, vs.concurrentUploadLimit)
+					err := fmt.Errorf("reject because inflight upload data %d > %d, and wait timeout", inFlightUploadSize, vs.concurrentUploadLimit)
 					glog.V(1).Infof("too many requests: %v", err)
 					writeJsonError(w, r, http.StatusTooManyRequests, err)
 					return
 				}
-				glog.V(4).Infof("wait because inflight upload data %d > %d", vs.inFlightUploadDataSize, vs.concurrentUploadLimit)
-				vs.inFlightUploadDataLimitCond.Wait()
+
+				glog.V(4).Infof("wait because inflight upload data %d > %d", inFlightUploadSize, vs.concurrentUploadLimit)
+				<-vs.inFlightUploadWaitCh
 			}
-			vs.inFlightUploadDataLimitCond.L.Unlock()
 		}
+
 		atomic.AddInt64(&vs.inFlightUploadDataSize, contentLength)
 		defer func() {
 			atomic.AddInt64(&vs.inFlightUploadDataSize, -contentLength)
-			if vs.concurrentUploadLimit != 0 {
-				vs.inFlightUploadDataLimitCond.Signal()
+			select {
+			case vs.inFlightUploadWaitCh <- struct{}{}:
+			default:
```

Comment:
with the default clause, this means the `vs.inFlightUploadWaitCh <- struct{}{}` may not actually happen if blocked?

Reply:
Yes, just like `cond.Signal()`